### PR TITLE
feat: DFlash (block-parallel) draft model training

### PR DESCRIPTION
**Pull request by zhubohao911**
## Objective

Implement DFlash block-parallel draft model training in TorchSpec, enabling disaggregated online training with SGLang inference backend.

DFlash predicts 16-token blocks in parallel using dual-source KV attention, achieving ~5x fewer forward passes per training step compared to Eagle3's autoregressive approach.

Reference: [z-lab/dflash](https://github.com/z-lab/dflash), [SpecForge DFlash PR](https://github.com/sgl-project/SpecForge/pull/415)

### DFlash vs Eagle3

| | DFlash | Eagle3 |
|---|--------|--------|
| Loss | CE (cross-entropy) | KL (forward KL) |
| Prediction | Block-parallel (block_size=16) | Autoregressive (ttt_length=7) |
| Target layers | 5 layers | 3 layers |
| Mask | Block-causal (FlexAttention) | Causal |
| Context | Dual-source KV (W_proj) | Input fusion (fc layer) |
| Forward passes per step | 1 | 7 (sequential) |

### Core Implementation (new)

| File | Description |
|------|-------------|
| `torchspec/models/draft/dflash.py` | `DFlashDraftModel` — 5-layer transformer with dual-source KV attention, GQA, shared embedding + LM head |
| `torchspec/models/dflash.py` | `DFlashModel` — training wrapper with anchor sampling, block-causal mask (FlexAttention), CE loss with exponential decay |
| `torchspec/training/dflash_trainer.py` | `DFlashTrainer` — FSDP2 integration, WSD LR scheduler, frozen LM head initialization |
| `torchspec/config/dflash_draft_config.json` | Default draft model config for Qwen3-8B |
| `configs/sglang_qwen3_8b_dflash.yaml` | Multi-GPU training config (SGLang + FSDP) |

### Tests

| File | Description |
|------|-------------|
| `tests/test_dflash.py` | 67 tests — config, architecture, anchor sampling, block-causal mask, forward pass, loss, accuracy |

### Tooling

| File | Description |
|------|-------------|
| `scripts/tools/extract_dflash_checkpoint.py` | FSDP distributed checkpoint → single .pt extraction |
| `scripts/tools/prepare_perfectblend.py` | PerfectBlend dataset download and normalization |
| `docs/dflash/README.md` | Architecture overview, training guide, and validation results |

### Modified Files

| File | Change |
|------|--------|
| `torchspec/models/target/eagle3_target_model.py` | Generalized `set_aux_hidden_states_layers()` for N layers (was hardcoded to 3) |
| `torchspec/models/ops/flex_attention.py` | Increased recompile limit, inductor GEMM backend config |
| `torchspec/training/trainer.py` | Async data prefetching, gradient sync for micro-batch accumulation |
| `torchspec/training/data_fetcher.py` | `PrefetchedDataFetcher`, `min_loss_tokens` filtering, clone safety for batch>1 |
| `torchspec/training/optimizer.py` | Added `min_lr` parameter to LR scheduler |
| `torchspec/training/eagle3_trainer.py` | Plumbed `weight_decay`, `min_lr`, WSD params |
| `torchspec/config/train_config.py` | DFlash-specific config fields, `min_loss_tokens`, `shuffle_dataset`, `max_checkpoints` |
| `torchspec/controller/loop.py` | Checkpoint rotation (`max_checkpoints`), verbose timing |
| `torchspec/training/trainer_actor.py` | Config-based trainer dispatch (`DFlashConfig` → `DFlashTrainer`) |
| `torchspec/train_entry.py` | Auto-set aux layer IDs for DFlash, inductor GEMM fix |
| `torchspec/inference/engine/__init__.py` | Lazy SGLang/vLLM imports (HF-only training support) |

## Validation Results

Best model (P2-WSD): 800K PerfectBlend, 3 epochs, WSD LR schedule, 8x H100.

| Dataset | Our τ | z-lab τ | E2E Speedup |
|---------|-------|---------|-------------|
| gsm8k | 3.89 | 3.38 | 2.47x |
| math500 | 4.19 | 4.61 | 2.80x |
| aime24 | 3.98 | 4.12 | 2.60x |
| aime25 | 3.69 | 4.07 | 2.42x |
| humaneval | 4.30 | — | 2.68x |
| livecodebench | 4.72 | — | 2.96x |

**Math avg τ = 3.94** (2.7% gap to z-lab's 4.05). Decode-only speedup reached **3.02x** on livecodebench.

## Usage

```bash
# Prepare data
python scripts/tools/prepare_perfectblend.py \
  --output data/perfectblend_50k.jsonl --sample-size 50000

# Train (4x H100: 1 inference + 3 training FSDP)
python -m torchspec.train_entry \
  --config configs/sglang_qwen3_8b_dflash.yaml \
  dataset.train_data_path=data/perfectblend_50k.jsonl \
  output_dir=./outputs/qwen3-8b-dflash

# Extract checkpoint
python scripts/tools/extract_dflash_checkpoint.py \
  --checkpoint_dir outputs/qwen3-8b-dflash/checkpoints/iter_NNNNNNN \
  --output dflash_draft.pt
```

## References

- DFlash paper & code: [z-lab/dflash](https://github.com/z-lab/dflash)
- SpecForge DFlash PR: https://github.com/sgl-project/SpecForge/pull/415

### CODE REVIEW

**Comment by cicirori** [comment]
Thanks for the thorough work! I happened to finish writing, and after publishing I came across this PR and made some changes based on it.  https://github.com/torchspec-project/TorchSpec/pull/71

**Comment by zhubohao911** [comment]
> Thanks for the thorough work! I happened to finish writing, and after publishing I came across this PR and made some changes based on it. #71

Hi cicirori

Can we merge this PR first? I believe @yubofredwang is reviewing this one.

**Comment by cicirori** [comment]
> > Thanks for the thorough work! I happened to finish writing, and after publishing I came across this PR and made some changes based on it. #71
> 
> Hi cicirori
> 
> Can we merge this PR first? I believe @yubofredwang is reviewing this one.

sure, linter failed. can you fix that?

**Review by cicirori — APPROVED** [review]

**Comment by zhubohao911** [comment]
> > > Thanks for the thorough work! I happened to finish writing, and after publishing I came across this PR and made some changes based on it. #71
> > 
> > 
> > Hi cicirori
> > Can we merge this PR first? I believe @yubofredwang is reviewing this one.
> 
> sure, linter failed. can you fix that?

Thank you so much!
